feat: Add ConnectionDefinition support to registration utilities

HeerakKashyap opened the pull request:
This PR fixes the issue where ConnectionDefinition entities weren't being imported properly - they just got ignored during model import. I added the missing connections.meshery.io case to FindEntityType in utils/utils.go, which was the main blocker. Also made getEntity public as GetEntity in models/registration/utils.go for testing purposes. Updated PackagingUnit in models/registration/register.go to include connections and added the registration logic. Added connection file handling in directory processing in models/registration/dir.go. Finally, added connection support to the database schema in models/meshmodel/registry/registry.go.

I tested this with a sample ConnectionDefinition JSON and it works correctly - FindEntityType identifies it as a connection, GetEntity parses it properly, and the whole pipeline works end-to-end. Before this fix, connection definitions were being silently ignored during model import. Now the registry can handle all entity types: components, relationships, and connections.

related to https://github.com/meshery/meshkit/issues/762
Notes for Reviewers

[Signed commits](https://github.com/meshery/meshkit/pull/CONTRIBUTING.md)
 Yes, I signed my commits.

vr-varad commented:
Thank you @HeerakKashyap  for your contribution!!
If this work is complete, this is a good item to add to the weekly [Meshery Development meeting agenda](https://docs.google.com/document/d/15IEwtTxPOkQXN2r_1LzGXuAv2mIUlmbdtmxpyxXF-KI/edit). You can add this item in the doc, attend, and present it. Meeting details at https://meet.layer5.io/meshery.

HeerakKashyap commented:
> Thank you @HeerakKashyap for your contribution!! If this work is complete, this is a good item to add to the weekly [Meshery Development meeting agenda](https://docs.google.com/document/d/15IEwtTxPOkQXN2r_1LzGXuAv2mIUlmbdtmxpyxXF-KI/edit). You can add this item in the doc, attend, and present it. Meeting details at https://meet.layer5.io/meshery.



> Thank you @HeerakKashyap for your contribution!! If this work is complete, this is a good item to add to the weekly [Meshery Development meeting agenda](https://docs.google.com/document/d/15IEwtTxPOkQXN2r_1LzGXuAv2mIUlmbdtmxpyxXF-KI/edit). You can add this item in the doc, attend, and present it. Meeting details at https://meet.layer5.io/meshery.

Yeah I'll let u know as soon as I am done with solving the bugs

HeerakKashyap commented:
Hey @n2h9 @leecalcote, I've updated the PR with all the fixes for the issues we were seeing. The description now explains what was broken and how I fixed it. Would appreciate a review when you have time!

n2h9 commented on utils/utils.go:
```diff
@@ -112,7 +112,7 @@ func DownloadFile(filepath string, url string) error {
 		return err
 	}
-	defer resp.Body.Close()
+	defer SafeClose(resp.Body)
```
I think in this file will be enough to just update switch case to handle case "connections.meshery.io" and updates to close are not related to the issue. If you think they improve the code quality, could you please create a new pr with it :innocent: .

HeerakKashyap replied:
yeah sure...

n2h9 commented on models/registration/utils.go:
```diff
@@ -4,48 +4,59 @@ import (
+
 // TODO: refactor this and use CUE
-func getEntity(byt []byte) (et entity.Entity, _ error) {
+func GetEntity(byt []byte) (et entity.Entity, _ error) {
```
I do not see any place where you use it ourtside of the package :woman_shrugging: let's probably update it back to be unexported?

HeerakKashyap replied:
done!

n2h9 commented on models/registration/utils.go:
```diff
@@ -4,48 +4,59 @@ import (
 )
 
+// ConnectionSchemaVersion is the schema version for connection definitions
+const ConnectionSchemaVersion = "connections.meshery.io/v1beta1"
```
please also define this in the schema package :innocent:

HeerakKashyap commented on models/meshmodel/core/policies/rego_policy_relationship.go:
```diff
@@ -8,12 +8,12 @@ import (
-	"github.com/open-policy-agent/opa/storage"
-	"github.com/open-policy-agent/opa/storage/inmem"
-	"github.com/open-policy-agent/opa/topdown/print"
+	"github.com/open-policy-agent/opa/rego"          //nolint:staticcheck // SA1019: deprecated package, migration to v1 API requires extensive changes
```
> Do you think that this pr: #789 fixes lint issues in opa package ?
> 
> I encourage you to go and leave your thoughts on the changes in this pr 🤗 .

Thanks for pointing this out @n2h9! Looking at PR #789, this is exactly what we needed - it's doing a proper OPA migration from v1.0.1 to v1.7.1 with clean API refactoring instead of just slapping on nolint comments. The Rego code looks much cleaner now with the new API patterns, and they're updating all the other dependencies too. This is way better than what we were trying to do in our PR. Since this PR is specifically about dependency modernization, it makes total sense to let it handle all the OPA and ORAS stuff while we keep our ConnectionDefinition PR focused on just that functionality. Much cleaner approach - thanks for letting me know... 😄

HeerakKashyap commented:
> Hey hey, @HeerakKashyap hello 👋 😇
> 
> You need to take connection definition from the schema repo and not define it in meshkit. Similar how f.e. in [getEntity](https://github.com/meshery/meshkit/blob/545d2908fc2f2062ee1285133feb7febc9077896/models/registration/utils.go): it is done for models, components and relationship.

Thanks for the feedback @n2h9! I looked into using the existing `Connection` struct from the schemas repo, but I found that it doesn't implement the `entity.Entity` interface that our registration system requires. The `Connection` struct's `Create` method signature is different from what our `Entity` interface expects.

For now, I've kept the `ConnectionDefinition` in MeshKit to get the registration functionality working. The proper long-term solution would be to create a `ConnectionDefinition` struct in the schemas repo that implements the `Entity` interface, similar to how `ComponentDefinition` and `ModelDefinition` work.

This way we can move forward with the registration system while maintaining the correct architectural pattern. Would you like me to create a separate PR to the schemas repo to add the proper `ConnectionDefinition` there?

n2h9 commented:
> > Hey hey, @HeerakKashyap hello 👋 😇
> > You need to take connection definition from the schema repo and not define it in meshkit. Similar how f.e. in [getEntity](https://github.com/meshery/meshkit/blob/545d2908fc2f2062ee1285133feb7febc9077896/models/registration/utils.go): it is done for models, components and relationship.
> 
> Thanks for the feedback @n2h9! I looked into using the existing `Connection` struct from the schemas repo, but I found that it doesn't implement the `entity.Entity` interface that our registration system requires. The `Connection` struct's `Create` method signature is different from what our `Entity` interface expects.
> 
> For now, I've kept the `ConnectionDefinition` in MeshKit to get the registration functionality working. The proper long-term solution would be to create a `ConnectionDefinition` struct in the schemas repo that implements the `Entity` interface, similar to how `ComponentDefinition` and `ModelDefinition` work.
> 
> This way we can move forward with the registration system while maintaining the correct architectural pattern. Would you like me to create a separate PR to the schemas repo to add the proper `ConnectionDefinition` there?

I think we need to add missing functions to connection in [connection_helper.go](https://github.com/meshery/schemas/blob/master/models/v1beta1/connection/connection_helper.go) to make it implement required interface. Use Component and Model entities (they also have xxx_helper.go file in schema repo. 
:innocent: 

thank you :bowing_woman:

HeerakKashyap commented:
or maybe @leecalcote could help us give their opinion reagarding this

> > > > Hey hey, @HeerakKashyap hello 👋 😇
> > > > You need to take connection definition from the schema repo and not define it in meshkit. Similar how f.e. in [getEntity](https://github.com/meshery/meshkit/blob/545d2908fc2f2062ee1285133feb7febc9077896/models/registration/utils.go): it is done for models, components and relationship.
> > > 
> > > 
> > > Thanks for the feedback @n2h9! I looked into using the existing `Connection` struct from the schemas repo, but I found that it doesn't implement the `entity.Entity` interface that our registration system requires. The `Connection` struct's `Create` method signature is different from what our `Entity` interface expects.
> > > For now, I've kept the `ConnectionDefinition` in MeshKit to get the registration functionality working. The proper long-term solution would be to create a `ConnectionDefinition` struct in the schemas repo that implements the `Entity` interface, similar to how `ComponentDefinition` and `ModelDefinition` work.
> > > This way we can move forward with the registration system while maintaining the correct architectural pattern. Would you like me to create a separate PR to the schemas repo to add the proper `ConnectionDefinition` there?
> > 
> > 
> > I think we need to add missing functions to connection in [connection_helper.go](https://github.com/meshery/schemas/blob/master/models/v1beta1/connection/connection_helper.go) to make it implement required interface. Use Component and Model entities (they also have xxx_helper.go file in schema repo. 😇
> > thank you 🙇‍♀️
> 
> Hey @n2h9! Thanks for the feedback! 👋 😅
> 
> I totally get where you're coming from - we should definitely use the schemas repo definitions when possible. But I ran into a bit of a problem when I tried to add those Entity methods to the existing `Connection` struct.
> 
> Here's what I found: the `Connection` struct from the schemas repo is actually playing a different role in our system. It's acting as a **registrant** (like a host/connection that registers other entities), not as an entity definition itself.
> 
> If you look at `RegistryManager.RegisterEntity()`, you'll see that `Connection` is the first parameter - it's the thing doing the registering, not the thing being registered. Adding `Type()`, `GetID()`, and `GetEntityDetail()` methods to it would kinda break that role and create some confusion about what it's supposed to do.

or maybe @leecalcote could help us give their opinion regarding this... 😊🙇

leecalcote commented:
> @n2h9 kindly let me know if the changes I did works for us 🙇

@HeerakKashyap, test your changes and provide confirmation of those test results.

HeerakKashyap commented:
@leecalcote I have tested locally - it was working fine. I'll upload the test results shortly to confirm it 👍

HeerakKashyap commented:
> Could you please also remove no-lint around opa packages import, there is a separate pr which fixes the opa packages lint errors

So should I leave it for now as it is...?

n2h9 commented:
> > Hey hey @HeerakKashyap hello 👋 😇
> > 
> > * you need to take Connection struct from schema, [here](https://github.com/meshery/schemas/blob/master/models/v1beta1/connection/connection.go).
> > * you do not need to create ConnectionDefinition wrapper around it.
> > * if you need to enhance Connection  struct with functions, you can do it in [connection_helper.go](https://github.com/meshery/schemas/blob/master/models/v1beta1/connection/connection_helper.go)
> > * You need to take this value `const ConnectionSchemaVersion = "connections.meshery.io/v1beta1"` from [here](https://github.com/meshery/schemas/blob/master/models/v1beta1/const.go#L7);
> > 
> > Could you please check how it is right now done for model, component and relationship?
> 
> I understand we don't want a ConnectionDefinition wrapper, but I'm facing a technical constraint: The Issue: The Connection struct from schemas has a Type field (in connection.go under meshkit/models/meshmodel/core/v1beta1), which conflicts with the Entity interface's required Type() method. This prevents the Connection struct from directly implementing the Entity interface.I Used Connection struct from schemas, Used ConnectionSchemaVersion from schemas, Couldn't add Entity methods to Connection due to field/method name conflict. How should I handle this? The registration system expects entities that implement the Entity interface, but the Connection struct can't implement it directly due to the Type field conflict.

Hey hey @HeerakKashyap hello :wave: :hugs: ! 

Indeed there is a conflict :cry: , could be a blocker.

We want to achieve is that the result PackagingUnit will contain the list of `Connection` entities from schema. 
You may wish to think about how to re-arrange the code in meshkit to achieve this result in efficient way.
F.e. rename the method name of an interface or the name of a Connection property (probably interface method will be easier to rename). Probably there is some easier and  smarter way to achieve this. 

Thank you :bowing_woman:

n2h9 commented:
Hey hey @HeerakKashyap hello :wave: :hugs: ! 

You are validating data over the connectionDefinition, which is a wrapper over connection entity. 
We must validate directly connection entity, because it is supposed to receive connection entity (not a wrapper) inside a model folder / archive when we call to model import.

n2h9 commented:
> Hey @n2h9, Let me know if this works...

hey hey Heerak, hello :wave: . Thank you, validation over the connection model looks good  :slightly_smiling_face: ! 

Couple points:
- we do not need to define connection json schema in meshkit `schemas/connections/connectionDefinition.json` ;
- we need to add Connection to resulting PackagingUnit ;

Thank you :hugs:

HeerakKashyap commented:
> > Hey @n2h9, Let me know if this works...
> 
> hey hey Heerak, hello 👋 . Thank you, validation over the connection model looks good 🙂 !
> 
> Couple points:
> 
> * we do not need to define connection json schema in meshkit `schemas/connections/connectionDefinition.json` ;
> * we need to add Connection to resulting PackagingUnit ;
> 
> Thank you 🤗

okay, i'll fix these..
